File: 0443-string-compression/0443-string-compression.js
/**
 * @param {character[]} chars
 * @return {number}
 */
var compress = function(chars) {
    let i = 0;  // Read pointer
    let res = 0;  // Write pointer (new length of compressed array)

    while (i < chars.length) {
        let groupLength = 1;  // Start by assuming a group of one character
        
        // Count how many consecutive identical characters there are
        while (i + groupLength < chars.length && chars[i + groupLength] === chars[i]) {
            groupLength++;
        }
        
        // Write the character to the result position
        chars[res++] = chars[i];

        // If the group length is greater than 1, write the count as characters
        if (groupLength > 1) {
            let countStr = String(groupLength);  // Convert the group length to a string
            for (let c of countStr) {
                chars[res++] = c;  // Write each digit of the count
            }
        }
        
        i += groupLength;  // Move the read pointer to the next group of characters
    }
    
    return res;  // Return the new length of the compressed array
};

File: 0443-string-compression/0443-string-compression.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        i = 0  # Pointer to read through the array
        res = 0  # Pointer to write the compressed result
        
        # Iterate through the entire chars array
        while i < len(chars):
            group_length = 1  # Start with a single character in the group

            # Count consecutive characters that are the same
            while (i + group_length < len(chars)
                   and chars[i + group_length] == chars[i]):
                group_length += 1  # Increment group length if characters are equal

            chars[res] = chars[i]  # Write the current character at the res pointer
            res += 1  # Move to the next position for writing

            # If the group length is greater than 1, write the count
            if group_length > 1:
                str_repr = str(group_length)  # Convert group length to string
                chars[res:res+len(str_repr)] = list(str_repr)  # Write count as individual digits
                res += len(str_repr)  # Update res to move past the written digits

            # Move i to the start of the next group
            i += group_length

        return res  # array
